server/src/main/java/org/opensearch/common/cache/tier/TieredSpilloverCache.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.tier;
+
+import org.opensearch.common.cache.ICache;
+import org.opensearch.common.cache.LoadAwareCacheLoader;
+import org.opensearch.common.cache.RemovalReason;
+import org.opensearch.common.cache.store.StoreAwareCache;
+import org.opensearch.common.cache.store.StoreAwareCacheRemovalNotification;
+import org.opensearch.common.cache.store.StoreAwareCacheValue;
+import org.opensearch.common.cache.store.builders.StoreAwareCacheBuilder;
+import org.opensearch.common.cache.store.enums.CacheStoreType;
+import org.opensearch.common.cache.store.listeners.StoreAwareCacheEventListener;
+import org.opensearch.common.util.concurrent.ReleasableLock;
+import org.opensearch.common.util.iterable.Iterables;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+/**
+ * This cache spillover the evicted items from heap tier to disk tier. All the new items are first cached on heap
+ * and the items evicted from on heap cache are moved to disk based cache. If disk based cache also gets full,
+ * then items are eventually evicted from it and removed which will result in cache miss.
+ *
+ * @param <K> Type of key
+ * @param <V> Type of value
+ *
+ * @opensearch.experimental
+ */
+public class TieredSpilloverCache<K, V> implements TieredCache<K, V>, StoreAwareCacheEventListener<K, V> {
+
+    // TODO: Remove optional when diskCache implementation is integrated.
+    private final Optional<StoreAwareCache<K, V>> onDiskCache;
+    private final StoreAwareCache<K, V> onHeapCache;
+    private final StoreAwareCacheEventListener<K, V> listener;
+    ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    ReleasableLock readLock = new ReleasableLock(readWriteLock.readLock());
+    ReleasableLock writeLock = new ReleasableLock(readWriteLock.writeLock());
```

Comment:
Looks slightly complex where spillover is orchestrated at multiple layers making the locks hard to maintain over time. See if we could get the tiered spillover atomic transition at a central place

Reply:
For now, I have created an issue for this [here](https://github.com/opensearch-project/OpenSearch/issues/11802)
Considering the spillover essentially is triggered from a removal listener from underlying onHeap cache implementation, it was getting difficult to simply this.

Let me know if it would be okay to refactor this later if possible.

Reply:
Makes sense